appinventor/blocklyeditor/src/blocklyeditor.js

```diff
@@ -514,79 +1085,128 @@ Blockly.ai_inject = function(container, workspace) {
-  Blockly.WidgetDiv.createDom();
-  Blockly.Tooltip.createDom();
   workspace.drawer_ = new Blockly.Drawer(workspace, { scrollbars: true });
-  workspace.flyout_ = workspace.drawer_.flyout_;
-  var flydown = new Blockly.Flydown(new Blockly.Options({scrollbars: false}));
-  // ***** [lyn, 10/05/2013] NEED TO WORRY ABOUT MULTIPLE BLOCKLIES! *****
-  workspace.flydown_ = flydown;
-  Blockly.utils.insertAfter_(flydown.createDom('g'), workspace.svgBubbleCanvas_);
-  flydown.init(workspace);
-  flydown.autoClose = true; // Flydown closes after selecting a block
+  workspace.flyout_ = workspace.getFlyout();
   workspace.addWarningIndicator();
   workspace.addBackpack();
-  Blockly.init_(workspace);
-  workspace.markFocused();
-  Blockly.bindEvent_(svg, 'focus', workspace, workspace.markFocused);
-  workspace.resize();
+  Blockly.browserEvents.bind(workspace.svgGroup_, 'focus', workspace, workspace.markFocused);
   // Hide scrollbars by default (otherwise ghost rectangles intercept mouse events)
   workspace.flyout_.scrollbar_ && workspace.flyout_.scrollbar_.setContainerVisible(false);
   workspace.backpack_.flyout_.scrollbar_ && workspace.backpack_.flyout_.scrollbar_.setContainerVisible(false);
   workspace.flydown_.scrollbar_ && workspace.flydown_.scrollbar_.setContainerVisible(false);
   // Render blocks created prior to the workspace being rendered.
-  workspace.rendered = true;
-  var blocks = workspace.getAllBlocks();
-
-  /**
-   * Creates a new helper function to render a comment set to visible but deferred during workspace
-   * generation.
-   * @param {!Blockly.Comment} comment The Blockly Comment object to be made visible.
-   * @returns {Function}
-   */
-  function commentRenderer(comment) {
-    return function() {
-      comment.setVisible(comment.visible);
-    }
-  }
+  workspace.injecting = false;
+  workspace.injected = true;
+  workspace.notYetRendered = true;
+  return workspace;
+};
 
-  for (var i = blocks.length - 1; i >= 0; i--) {
-    var block = blocks[i];
-    block.initSvg();
-    block.rendered = true;
-    if (block.disabled && block.updateDisabled) {
-      block.updateDisabled();
-    }
-    if (!isNaN(block.x) && !isNaN(block.y)) {
-      var xy = block.getRelativeToSurfaceXY();
-      block.getSvgRoot().setAttribute('transform',
-        'translate(' + block.x + ',' + block.y + ')');
-      block.moveConnections_(block.x - xy.x, block.y - xy.y);
-    }
-    if (block.comment && block.comment.visible && block.comment.setVisible) {
-      setTimeout(commentRenderer(block.comment), 1);
-    }
+/**
+ * Inject a previously constructed workspace into the designated
+ * container. This implementation is adapted from Blockly's
+ * implementation and is required due to the fact that browsers such as
+ * Firefox do not initialize SVG elements unless they are visible.
+ *
+ * @param {!Element|string} container
+ * @param {!Blockly.WorkspaceSvg} workspace
+ */
+Blockly.ai_inject = function(container, workspace) {
```

Comment:
We may just want to call this `AI.inject` to mirror `Blockly.inject` rather than attach an App Inventor specific function to the Blockly workspace.

Reply:
Agreed!